Textarea Field, Multiline Block

acbart opened the pull request:
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

https://github.com/google/blockly/pull/30

### Proposed Changes

* Creates TextArea field that descends from TextInput
* Creates a Multiline Text Block
* Provides generators for new multiline text block
* Add multiline_quote_ function to all languages to properly handle creating multiline strings.
* Creates Pilcrow text mixin

### Reason for Changes

A number of projects (e.g., BlockPy) make use of multiline text blocks. Common purposes include long strings, comments, and raw code blocks. This pull request makes the textarea another possible field type.

### Test Coverage

I added the block to the playground and ensured that the code could be generated for all the languages.

Tested on:
* Desktop Chrome

### Additional Information

* I'm not sure I handled every possible multiline edge case for Dart, Lua, PHP. I'm not as practiced with those languages, and it seems like it'd be possible to circumvent the text replacing easily. But then again, I'm not sure how much effort fighting those cases is worth.
* I used the Wikimedia commons for the Pilcrow image. Let me know if this is acceptable - seems like they have a permissive license.

acbart commented:
I signed it!

BeksOmega commented on blocks/text.js:
```diff
@@ -647,6 +664,80 @@ Blockly.Constants.Text.TEXT_QUOTES_EXTENSION = function() {
+ * @package
+ * @readonly
+ */
+Blockly.Constants.Text.PILCROW_IMAGE_MIXIN = {
```
Instead of having this as a mixin I think it's better to just define an image field on the block. This will make the src path static, but I'm planning on adding dynamic media path support to the JSON at some point, so it will only be temporary.

BeksOmega commented on tests/playground.html:
```diff
@@ -660,6 +660,7 @@ <h1>Blockly Playground</h1>
     </category>
     <category name="Text" categorystyle="text_category">
       <block type="text"></block>
+      <block type="text_multiline"></block>
```
Please also add this to the typed variables toolbox, and the multi-playground categories toolbox.

acbart commented:
Wow, thanks for the in-depth feedback! I'll try to get this all responded to this weekend.

For #1, I see what you mean about the round-trip. Seems like an issue with a new line being added. I'm not quite sure what part of the XML hijinx are causing that, so using the multi-part value may be an expedient way to bypass the issue.

Looking at #2, I did view it as a single paragraph. In particular, I use this field in BlockPy for multiline strings and raw code blocks, so new lines are fairly precise. Text wrapping wouldn't be my first choice, personally, but if that was more proper for Blockly I wouldn't be opposed.

Regarding #3, should I wait till the PR is accepted to create the issue?

BeksOmega commented:
> Looking at question 2, I did view it as a single paragraph. In particular, I use this field in BlockPy for multiline strings and raw code blocks, so new lines are fairly precise. Text wrapping wouldn't be my first choice, personally, but if that was more proper for Blockly I wouldn't be opposed.

Ah ok I see what you're going for! I'm not sure what the most common use case for a multiline text field is, so I'll defer to the team's judgment on this one.
 
> Regarding question 3, should I wait till the PR is accepted to create the issue?

Hmmm... Now that you mention it I think you're right to wait on it. Wouldn't want anyone to try and jump on it and get confused.

acbart commented on core/field_textarea.js:
```diff
@@ -0,0 +1,217 @@
+ * @return {!HTMLInputElement} The newly created text input editor.
+ * @private
+ */
+Blockly.FieldTextArea.prototype.widgetCreate_ = function() {
```
I don't think this will work, unfortunately. FieldTextArea creates a `textarea` element, while FieldTextInput creates an `input` element. I could refactor widgetCreate_ to create its element based on a field, but that feels outside my scope.

acbart commented:
* I fixed an issue where zooming would mess up the scale of the textarea.
* I'm not 100% convinced I have the textarea properly aligned on the rendered textelement, but I think it's close enough to be effective.
* In RTL mode, the text is still left aligned. I'm not sure what the precise proper behavior should be.

acbart commented:
Okay, I think those commits resolve the issues raised so far. Let me know if there's more before it can be accepted.

amber-cd commented:
Re: text wrapping, would it be appropriate to have the field accept some type of "wrap text" parameter to allow for both use cases? I hate to ask you to do more work, particularly as a volunteer who implemented an oft-requested feature (which looks great, by the way), but it seems like having some type of configuration parameter in the block constructor which controls text wrapping might be worthwhile.

acbart commented:
> Re: text wrapping, would it be appropriate to have the field accept some type of "wrap text" parameter to allow for both use cases?

The parameter would be easy to add, but what specific functionality would it trigger? Instead of triggering ellipsis, it just wraps onto a new line?

amber-cd commented:
More or less. My thought would be specifically for something like, say, a "send email" block where you can send an email with the text and don't want to manually add in line breaks except when making new paragraphs, but want to be able to see the full text of each paragraph.

acbart commented on core/field_textarea.js:
```diff
@@ -0,0 +1,238 @@
+  div.style.fontSize = fontSize;
+  
+  var scale = this.sourceBlock_.workspace.scale;
+  div.style.lineHeight = (20 * scale) + 'px';
```
Ah, I see now. I was dying inside writing out all these magic numbers, since I wasn't sure what fell under "general rendering weirdness" vs not. I'm glad to hear that there's a plan for these.

rachel-fenichel replied:
With respect to "general rendering weirdness", we've been rewriting rendering completely, with a lot of success.  The meat of the new rendering code is [here](https://github.com/google/blockly/tree/render/collab/core/block_rendering_rewrite) if you want to check it out.  TLDR is that everything on a field gets an explicit size before we try to render it, and fields are no longer responsible for horizontal and vertical padding.  (And there's a backwards compatibility plan--more on that to come).

acbart commented:
Haven't done anything with the text-wrapping (perhaps a separate issue, just to move this along?), but I renamed the field to MultilineText and got rid of the generated language files (oops). I'll be interested in hearing your more detailed review.

rachel-fenichel commented:
I agree that the XML question can be a separate issue, since it sounds like there are a few different use cases.

On the generators: multiline strings are ES6, so handle it by adding together strings (or using an array and join) instead.  Otherwise we get into a case where we can run in a browser (IE) and generate code that can't be executed in that same browser.

NeilFraser commented:
A patch file for fixing some style nits:
[0001-Fix-some-style-guide-nits.txt](https://github.com/google/blockly/files/3331784/0001-Fix-some-style-guide-nits.txt)

I'm not convinced that the XML format currently being is the best option.  Can't we encode the string with the usual escape characters?  I think we could make the XML for text and multiline text be compatible and interchangeable.  The XML format is one of the few things we don't get to change our minds on, so getting it right is important.

NeilFraser commented:
Formal suggestion for XML format.

The current text input field encodes the string "\<HTML\>" like this:
`<field name="TEXT">&lt;HTML&gt;</field>`

Ideally the multiline text input would use exactly the same format.  The problem is line feed characters.

* Using an actual line feed in the XML works, but makes manipulating the XML (e.g. pretty-printing it) very fragile.
* Using the string `'\n'` is not right since the literal string `'\n'` currently encodes as `'\n'` in a text input field.  Besides, `'\n'` is not an XML format.
* Using the string '&#10;' would seem to have no issues.

Thus, this pathological string:

```
Hello world
<HTML>
Literal \n string
Literal &#10; string
```

would encode as:

`<field name="TEXT">Hello world&#10;&lt;HTML&gt;&#10;Literal \n string&#10;Literal &amp;#10; string</field>`

The best part is that the &#10; encoding can be added to FieldTextInput, rather than FieldMultilineText.

NeilFraser commented:
> The best part is that the encoding can be added to FieldTextInput, rather than FieldMultilineText.

Correction: Field.to/fromXml would be where I'd suggest adding the `&#10;` encoder/decoder.

acbart commented:
Seems simple enough - do you want me to add this modification to this pull request? It'd basically just be:
````
fieldElement.textContent = this.getValue().replace('\n', '&#10;');
````

wouldn't it?

acbart commented:
I added the proposed fixes, by my understanding. Let me know if you had a different vision. Does seem simpler and more convenient - but given that I just modified Field, someone more familiar with the impacts should definitely check things over.

NeilFraser commented:
Your changes are exactly as I described.  And they don't work.  My fault.  :)

The issue is that we never explicitly encode or decode \<, \>, and \&.  That's handled automatically by the browser when one sets or gets `textContent`.  Thus when we set a string that contains `'&#10;'`, it automatically gets escaped to `'&amp;#10;'`.  As a result the current version of this PR will take the user text `'foo&#10;bar'`, encode it as `<field name="TEXT">foo&amp;#10;bar</field>`, and round trip it back as
```
foo
bar
```

An approach I've tested that works is to remove the `&#10;` from both Blockly.Field.prototype.to/fromXml, and instead edit Blockly.Xml.domToText:
```
Blockly.Xml.domToText = function(dom) {
  return Blockly.Xml.utils.domToText(dom).replace(/\n/g, '&#10;');
};
```

acbart commented:
So, I forgot to run `jsunit` before I pushed, but as you can see that seems to have broken the build. The issues seems to be that we're now encoding *all* newlines with `'&#10;'`, including the ones that are outside of the strings such as the newlines that are apparently put between XML nodes. Is this desirable? I'd need to update a number of unit tests, it seems.

RoboErikG commented:
Looks like it's two tests:
test_domToText
test_domToPrettyText

My recommendation (unless Neil has another idea) would be to fix test_domToText by doing the same substitution in the expected string and fix test_domToPrettyText by modifying Blockly.Xml.domToPrettyText to substituting newlines back into the blob so it's still pretty. ;)

NeilFraser commented:
Well that's interesting.  The playground's XML does not reflect this.  Give me a chance to fully understand what's going on here.

BTW, this change has the side-effect of improving the XML for multi-line comments.  Fortunately in a backwards-compatible manner.

NeilFraser commented:
This was fun.  I don't think this issue would ever occur outside of a unit test, since it requires uselessly round-tripping XML without ever using it in Blockly.  Also worth noting all the variants of this XML are perfectly valid and will work fine.  The current issue is purely cosmetic (though any difference will fail the unit test's equality operator).

Nevertheless, here's a replacement function that resolves this issue.  The use of a completely unreadable regular expression does not fill me with joy, however, if it fails there are no runtime consequences.

```
/**
 * Converts a DOM structure into plain text.
 * Currently the text format is fairly ugly: all one line with no whitespace,
 * unless the DOM itself has whitespace built-in.
 * @param {!Element} dom A tree of XML elements.
 * @return {string} Text representation.
 */
Blockly.Xml.domToText = function(dom) {
  var text = Blockly.Xml.utils.domToText(dom);
  // Replace line breaks in text content with '&#10;' to make them single line.
  // E.g. <foo>hello\nworld</foo> -> <foo>hello&#10;world</foo>
  // Do not replace line breaks between tags.
  // E.g. ...</foo>\n</bar> is unchanged.
  // Can't use global flag on regexp since backtracking is needed.
  var regexp = /(<[^/][^<]*>[^<]*)\n([^<]*<\/)/;
  var oldText;
  do {
    oldText = text;
    text = text.replace(regexp, '$1&#10;$2');
  } while (text != oldText);
  return text;
};
```

JSUnit tests pass, as does Mocha.

acbart commented:
I had been concerned that this additional regexing would negatively impact performance, but surprisingly it doesn't seem to have much effect on the timing runs I ran. I'll get this in and repush.

NeilFraser commented:
XML looks good to me.

acbart commented:
Where do we go from here? Do you need anything else from me, or does the review process just continue?

RoboErikG commented:
Whoops, sorry for the delay. I think this is mostly good to go (@rachel-fenichel to confirm), but we'll wait until late next week to merge as we're finalizing this quarter's release this.

rachel-fenichel commented:
Yup, it looks solid.  We'll cut the release candidate at the end of this week or start of next week, and merge this into develop after that cut.

It looks like there is one trivial merge conflict from Neil's recent XML changes; either you'll fix that or I will next week.
